app/assets/javascripts/discourse/app/controllers/preferences/interface.js

```diff
@@ -434,52 +476,110 @@ export default class InterfaceController extends Controller {
     this.setProperties({
       selectedDarkColorSchemeId: colorSchemeId,
-      previewingColorScheme: this.isViewingOwnProfile,
+      previewingColorScheme: this._shouldEnablePreview(true),
     });
 
     if (!this.isViewingOwnProfile) {
       return;
     }
 
-    if (colorSchemeId === -1) {
-      // load preview of regular scheme when dark scheme is disabled
-      loadColorSchemeStylesheet(this.selectedColorSchemeId, this.themeId, true);
-      this.session.set("darkModeAvailable", false);
-    } else {
-      loadColorSchemeStylesheet(colorSchemeId, this.themeId, true);
-      this.session.set("darkModeAvailable", true);
+    // only preview dark schemes when in dark mode
+    if (!this._isInDarkMode()) {
+      return;
     }
+
+    this._previewColorScheme(true);
+    this.session.set("darkModeAvailable", colorSchemeId !== -1);
   }
 
   @action
   selectColorMode(modeId) {
     this.selectedInterfaceColorModeId = modeId;
     this.set("previewingColorScheme", this.isViewingOwnProfile);
+
+    if (!this.isViewingOwnProfile) {
+      return;
+    }
+
+    this._applyInterfaceModePreview(modeId);
+    this._previewColorSchemeForMode(modeId);
+  }
+
+  _applyInterfaceModePreview(modeId) {
+    const modeActions = {
+      [INTERFACE_COLOR_MODES.AUTO]: () => this.interfaceColor.useAutoMode(),
+      [INTERFACE_COLOR_MODES.LIGHT]: () => this.interfaceColor.forceLightMode(),
+      [INTERFACE_COLOR_MODES.DARK]: () => this.interfaceColor.forceDarkMode(),
+    };
+
+    modeActions[modeId]?.();
+  }
+
+  _previewColorSchemeForMode(modeId) {
+    if (this._shouldShowPreviewForMode(modeId, false)) {
+      this._removePreviewStylesheet("dark");
+      this._previewColorScheme(false);
+    } else if (this._shouldShowPreviewForMode(modeId, true)) {
+      this._removePreviewStylesheet("light");
+      this._previewColorScheme(true);
+    }
+  }
+
+  _shouldShowPreviewForMode(modeId, isDark) {
+    const targetMode = isDark
+      ? INTERFACE_COLOR_MODES.DARK
+      : INTERFACE_COLOR_MODES.LIGHT;
+    const autoCondition = isDark
+      ? window.matchMedia("(prefers-color-scheme: dark)").matches
+      : !window.matchMedia("(prefers-color-scheme: dark)").matches;
+
+    return (
+      modeId === targetMode ||
+      (modeId === INTERFACE_COLOR_MODES.AUTO && autoCondition)
+    );
+  }
+
+  _removePreviewStylesheet(type) {
+    const selector =
+      type === "dark" ? "link#cs-preview-dark" : "link#cs-preview-light";
+    const stylesheet = document.querySelector(selector);
+    if (stylesheet) {
+      stylesheet.remove();
+    }
+  }
+
+  _previewColorScheme(isDark) {
+    const selectedId = isDark
+      ? this.selectedDarkColorSchemeId
+      : this.selectedColorSchemeId;
+    const colorSchemeId = this._resolveThemeDefaultColorScheme(
+      selectedId,
+      isDark
+    );
+    loadColorSchemeStylesheet(colorSchemeId, this.themeId, isDark);
+
+    if (!isDark) {
+      loadColorSchemeStylesheet(colorSchemeId, this.themeId, true);
+    }
```

Comment:
just writing it out to wrap my head around it 😅 

if the color scheme is dark:
* load the stylesheet with isDark `true`

if the color scheme is light:
* load the stylesheet with isDark `false`
* load the stylesheet with isDark `true`

if that's correct, maybe it would be more readable to have the code follow that? i.e.

```js
if (isDark) {
  load(…, true);
} else {
  load(…, true);
  load(…, false);
}
```